homeassistant/components/sensor/__init__.py

```diff
@@ -536,39 +605,49 @@ def state(self) -> Any:
                     report_issue,
                 )
                 return value
+        else:
+            numerical_value = value
+
+        if (
+            prec is not None
+            and (self._sensor_option_precision or not isinstance(value, int))
+            and (
+                native_unit_of_measurement == unit_of_measurement
+                or device_class not in UNIT_CONVERTERS
+            )
+        ):
+            value = f"{numerical_value:.{prec}f}"
```

Comment:
I think this should be moved down
```suggestion
        elif precision is not None and (
            self._sensor_option_precision or not isinstance(value, int)
        ):
            # No unit conversion, but precision is set
            value = f"{numerical_value:.{precision}f}"
```